### PR TITLE
kernel: Always set clock expiry with sync with timeout module

### PR DESCRIPTION
**Pull request by pdunaj**
System must not set the clock expiry via backdoor as it may
effect in unbound time drift of all scheduled timeouts.

Fixes: #11502

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

**Comment by pdunaj** [comment]
Tested on PCA20041 where original problem was observed. No issues seen apart from the fact that delayed works that are now finally being scheduled for execution impact the overall system performance.

It reminds me of the project I heard of in the past where developers commented out mutex code to get rid of deadlock. After this change the device was probably the fastest on the market... and was crashing randomly from time to time ;)

File: kernel/include/timeout_q.h [hunk 1]
```diff
@@ -48,6 +48,8 @@ static inline int _abort_thread_timeout(struct k_thread *thread)
 
 s32_t _get_next_timeout_expiry(void);
 
+void _set_timeout_expiry(s32_t ticks, bool idle);
```
Comment (andyross) [inline kernel/include/timeout_q.h]: Nitpick: Should be "z_set_timeout_expiry()" if it's not a static.  We're not allowed to use prepended underscores anymore.  Probably also wants documentation pointing out that this is the kernel API that wraps z_clock_set_timeout(), and that the driver API shouldn't be used outside the framework.

File: kernel/include/timeout_q.h [hunk 2]
```diff
@@ -57,6 +59,7 @@ s32_t z_timeout_remaining(struct _timeout *timeout);
 #define _add_thread_timeout(th, to) do {} while (0 && (void *)to && (void *)th)
 #define _abort_thread_timeout(t) (0)
 #define _get_next_timeout_expiry() (K_FOREVER)
+#define _set_timeout_expiry(t, i) do {} while (0)
```
Comment (andyross) [inline kernel/include/timeout_q.h]: Reminds me that I need to figure out the "right way" to do these noop fallbacks.  Empty function bodies were problematic to MISRA apparently, but I think I might be winning that fight.

**Review by andyross — CHANGES_REQUESTED** [review]
Code looks good with one minor naming issue.  Note that I have work in the pipe to remove the timeout setting operation from the idle loop entirely, but I can rebase on top of this.   I'd actually wrapped the z_clock_set_timeout() call in more or less the same way, so I'd be happy to see this go in instead.

Also, can you augment the commit message with a description of the race being fixed?  Something like "The read of the next timeout and setting it at the driver level need to be atomic, otherwise two threads can race if a sooner-expiring timeout is added after another thread has already decided to set a later one."

File: kernel/include/timeout_q.h (hunk 1 above)
Comment (pdunaj) [reply kernel/include/timeout_q.h]: Fixed name

**Review by andyross — APPROVED** [review]
Looks great, thanks.  Note that the CI failures seem to be straightforward unused variable warnings in oddball configurations; should be fixed, but surely easy.

**Comment by pdunaj** [comment]
I have run some tests today and last change works fine.

**Comment by pdunaj** [comment]
@pizi-nordic can you have a look?

**Review by pizi-nordic — APPROVED** [review]

**Comment by pdunaj** [comment]
@carlescufi , could you merge?

**Review by carlescufi — APPROVED** [review]

**Pull request merged**